compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/PromisedValue.kt

```diff
@@ -70,6 +77,30 @@ abstract class PromisedValue(val codegen: ExpressionCodegen, val type: Type, val
 
     val typeMapper: IrTypeMapper
         get() = codegen.typeMapper
+
+    private fun IrType.isJava() =
+        (this.classifierOrFail.owner as IrDeclaration).origin == IrDeclarationOrigin.IR_EXTERNAL_JAVA_DECLARATION_STUB
+
+    private fun IrType.isPackagePrivate(): Boolean {
+        return this.getClass()?.visibility?.delegate == JavaVisibilities.PackageVisibility
+    }
+
+    private fun IrType.inDifferentPackage(): Boolean =
+        this.getClass()!!.getPackageFragment()?.fqName != codegen.irFunction.getPackageFragment()?.fqName
+
+    private fun computeCoerceTarget(
+        isUpcastToJava: Boolean,
+        erasedTargetType: IrType,
+        default: Type
+    ): Type {
+        if (!isUpcastToJava && erasedTargetType.isJava() && erasedTargetType.isPackagePrivate() && erasedTargetType.inDifferentPackage()) {
```

Comment:
> `!isUpcastToJava && erasedTargetType.isJava()`

This is basically checking that the `isSubtypeOfClass` call at the call site returned `false` (because `isUpcastToJava` also contains the check for `erasedTargetType.isJava()`). So maybe the first parameter to this function should rather be something like `isSubtype: Boolean`?

Reply:
Done, and a few redundant conditional checks has been removed , it looks more readable now.